src/common/scheduler/scheduler.go

```diff
@@ -0,0 +1,258 @@
+package scheduler
+
+import "github.com/vmware/harbor/src/common/scheduler/policy"
+import "github.com/vmware/harbor/src/common/utils/log"
+
+import "errors"
+import "strings"
+import "fmt"
+import "reflect"
+import "time"
+
+const (
+	defaultQueueSize = 10
+
+	statSchedulePolicy   = "Schedule Policy"
+	statUnSchedulePolicy = "Unschedule Policy"
+	statTaskRun          = "Task Run"
+	statTaskComplete     = "Task Complete"
+	statTaskFail         = "Task Fail"
+)
+
+//StatItem is defined for the stat metrics.
+type StatItem struct {
+	//Metrics catalog
+	Type string
+
+	//The stat value
+	Value uint32
+
+	//Attach some other info
+	Attachment interface{}
+}
+
+//StatSummary is used to collect some metrics of scheduler.
+type StatSummary struct {
+	//Count of scheduled policy
+	PolicyCount uint32
+
+	//Total count of tasks
+	Tasks uint32
+
+	//Count of successfully complete tasks
+	CompletedTasks uint32
+
+	//Count of tasks with errors
+	TasksWithError uint32
+}
+
+//Configuration defines configuration of Scheduler.
+type Configuration struct {
+	QueueSize uint8
+}
+
+//Scheduler is designed for scheduling policies.
+type Scheduler struct {
+	//Related configuration options for scheduler.
+	config *Configuration
+
+	//Store to keep the references of scheduled policies.
+	policies Store
+
+	//Queue for accepting the scheduling polices.
+	scheduleQueue chan policy.Policy
+
+	//Queue for receiving policy unschedule request or complete signal.
+	unscheduleQueue chan string
+
+	//Channel for receiving stat metrics.
+	statChan chan *StatItem
+
+	//Channel for terminate scheduler damon.
+	terminateChan chan bool
+
+	//The stat metrics of scheduler.
+	stats *StatSummary
+
+	//To indicate whether scheduler is stopped or not
+	stopped bool
+}
+
+//DefaultScheduler is a default scheduler.
+var DefaultScheduler = NewScheduler(nil)
+
+//NewScheduler is constructor for creating a scheduler.
+func NewScheduler(config *Configuration) *Scheduler {
+	var qSize uint8 = defaultQueueSize
+	if config != nil && config.QueueSize > 0 {
+		qSize = config.QueueSize
+	}
+
+	sq := make(chan policy.Policy, qSize)
+	usq := make(chan string, qSize)
+	stChan := make(chan *StatItem, 4)
+	tc := make(chan bool, 2)
+
+	store := NewConcurrentStore(10)
+	return &Scheduler{
+		config:          config,
+		policies:        store,
+		scheduleQueue:   sq,
+		unscheduleQueue: usq,
+		statChan:        stChan,
+		terminateChan:   tc,
+		stats: &StatSummary{
+			PolicyCount:    0,
+			Tasks:          0,
+			CompletedTasks: 0,
+			TasksWithError: 0,
+		},
+		stopped: true,
+	}
+}
+
+//Start the scheduler damon.
+func (sch *Scheduler) Start() {
+	if !sch.stopped {
+		return
+	}
+	go func() {
+		defer func() {
```

Comment:
Tried, seems panic in the sub goroutine might not be caught by upper component. Besides double confirmation, it can also catch the panic errors from itself.